### PR TITLE
feat(datasource): remove deleted columns and update column type on metadata refresh

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Currently refreshing table metadata doesn't remove columns that aren't available on the table any longer. In addition, the React Table view doesn't update column types if they've changed. This PR:
- removes columns that are not present on table schema (both legacy and React).
- updates column type if changed (React; already working on legacy view).
- adds toasts for added, removed and update columns to (both legacy and React).
- Updates unit tests on React Jest specs to correspond to the new logic (skipped legacy view as I assume it's going to be deprecated).
- fix broken style on type pills on the table metadata column on React views (see screenshots).

### AFTER
image
image

### BEFORE
The type pills on the column tab were using an incorrect style, causing white foreground on white background.
image

### TEST PLAN
Local testing + new tests

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 1]
```diff
@@ -85,11 +74,65 @@ describe('DatasourceEditor', () => {
...
+        primary_key: false,
+      },
+      {
+        name: 'new_column',
+        type: 'VARCHAR(10)',
+        nullable: true,
+        default: '',
+        primary_key: false,
+      },
+    ];
```
Comment (villebro) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:100]: This is the correct schema coming back from the API (compare with `extraColumn` above).

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 2]
```diff
@@ -85,11 +74,65 @@ describe('DatasourceEditor', () => {
...
+        {
+          type: 'DATETIME',
+          description: null,
+          filterable: false,
+          verbose_name: null,
+          is_dttm: true,
+          expression: '',
+          groupby: false,
+          column_name: 'ds',
+        },
```
Comment (villebro) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:116]: This column is unchanged.

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 3]
```diff
@@ -85,11 +74,65 @@ describe('DatasourceEditor', () => {
...
+        {
+          type: 'VARCHAR(32)',
+          description: null,
+          filterable: true,
+          verbose_name: null,
+          is_dttm: false,
+          expression: '',
+          groupby: true,
+          column_name: 'gender',
+        },
```
Comment (villebro) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:126]: This type is updated from `VARCHAR(16)` to `VARCHAR(32)`

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 4]
```diff
@@ -85,11 +74,65 @@ describe('DatasourceEditor', () => {
...
+        expect.objectContaining({
+          column_name: 'new_column',
+          type: 'VARCHAR(10)',
+        }),
```
Comment (villebro) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:130]: This is a new column.

File: superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx [hunk 5]
```diff
@@ -85,11 +74,65 @@ describe('DatasourceEditor', () => {
...
+      ]),
+    );
+    expect(inst.state.databaseColumns).not.toEqual(
+      expect.arrayContaining([expect.objectContaining({ name: 'name' })]),
+    );
```
Comment (villebro) [inline superset-frontend/spec/javascripts/datasource/DatasourceEditor_spec.jsx:135]: This is one of the removed columns in the original metadata.

File: superset-frontend/src/datasource/DatasourceEditor.jsx [hunk 6]
```diff
@@ -289,29 +289,52 @@ export class DatasourceEditor extends React.PureComponent {
     this.validate(this.onChange);
   }
 
-  mergeColumns(cols) {
-    let { databaseColumns } = this.state;
-    let hasChanged;
-    const currentColNames = databaseColumns.map(col => col.column_name);
+  updateColumns(cols) {
```
Comment (villebro) [inline superset-frontend/src/datasource/DatasourceEditor.jsx:292]: Method was renamed to reflect that metadata is updated, not only merged to.

File: superset/connectors/sqla/models.py [hunk 7]
```diff
@@ -1247,35 +1254,45 @@ def fetch_metadata(self, commit: bool = True) -> None:
         any_date_col = None
         db_engine_spec = self.database.db_engine_spec
         db_dialect = self.database.get_dialect()
-        dbcols = (
-            db.session.query(TableColumn)
-            .filter(TableColumn.table == self)
-            .filter(or_(TableColumn.column_name == col.name for col in table_.columns))
```
Comment (villebro) [inline superset/connectors/sqla/models.py:1253]: I'm not sure what the intent here was, but I don't see any scenario where we want to bring in columns from another table with the same column names.

**Review by willbarrett — APPROVED** [review]
This looks like a good improvement/fix - I'd like to tag @lilykuang as well who's touched this code recently for a review.

**Review by lilykuang — APPROVED** [review]
LGTM 👍  thank you for the improvement

**Review by nytai — APPROVED** [review]
lgtm

**Pull request merged**